### PR TITLE
MDCT-2457, 2458, 2459: Restrict Report Access for State Users w/o Permissions

### PR DESCRIPTION
**Pull request by karla-vm**
### Description

Limiting report access and actions for state users that may not have permission for certain report types (MCPAR, MLR, NAAAR)

### Related ticket(s)

MDCT-2457
MDCT-2458
MDCT-2459

---
### How to test

Everything should work as expected for now

### Important updates

N/A

---
### Author checklist

- [ ] I have performed a self-review of my code
- [ ] I have added [thorough](https://bit.ly/3zPrxuZ) tests, if necessary
- [ ] I have updated relevant documentation, if necessary
---

_convert to a different template: [test → val](?expand=1&template=test-to-val-deployment.md)_ | _[val → prod](?expand=1&template=val-to-prod-deployment.md)_

### CODE REVIEW

File: services/ui-src/src/components/app/AppRoutes.tsx [hunk 1]
```diff
@@ -21,8 +21,11 @@ import { ScrollToTopComponent, useUser } from "utils";
 import { Fragment } from "react";
 
 export const AppRoutes = () => {
-  const { userIsAdmin } = useUser().user ?? {};
+  const { userIsAdmin, userReports } = useUser().user ?? {};
   const mlrReport = useFlags()?.mlrReport;
+  // determine if the user has access to specific reports
+  const hasMcpar = userReports?.includes("MCPAR") || userIsAdmin;
+  const hasMlr = userReports?.includes("MLR") || userIsAdmin;
```
Comment (braxex) [inline services/ui-src/src/components/app/AppRoutes.tsx:28]: Nit: Would maybe name these `hasMcparAccess` or similar. Alternatively could make a lookup object that has something like: 
```
userReportAccess = {
  MCPAR: userReports?.includes("MCPAR") || userIsAdmin,
  MLR: userReports?.includes("MLR") || userIsAdmin,
}
```
then access like:
```
userReportAccess["MCPAR"] ? whatever : whatever
```

File: services/ui-src/src/components/cards/TemplateCard.tsx [hunk 2]
```diff
@@ -67,7 +67,7 @@ export const TemplateCard = ({
               leftIcon={
                 <Image src={downloadIcon} alt="Download Icon" height="1.5rem" />
               }
-              isDisabled={isDisabled}
+              isDisabled={false}
```
Comment (braxex) [inline services/ui-src/src/components/cards/TemplateCard.tsx:70]: Question: Why is this one always false?

File: services/ui-src/src/components/forms/AdminDashSelector.tsx [hunk 3]
```diff
@@ -51,7 +53,7 @@ export const AdminDashSelector = ({ verbiage }: Props) => {
     selectedReport = selectedReport.replace("report-", "").toLowerCase();
     localStorage.setItem("selectedReportType", selectedReport);
 
-    if (userIsAdmin || userIsReadOnly) {
+    if (userCanReadOnly) {
```
Comment (braxex) [inline services/ui-src/src/components/forms/AdminDashSelector.tsx:56]: Comment: I'm on the fence with ones like this -- it might be more clear to keep it as the two explicitly noted groups.

Comment (braxex) [reply services/ui-src/src/components/forms/AdminDashSelector.tsx:56]: ^^ This applies to all similar spots where we collapse the user groups `userIsAdmin` and `userIsReadOnly` (which for the uninitiated are composed of [super admins (bo reps) and approvers] and [help desk and internal users] respectively) down to `userCanReadOnly`.

File: services/ui-src/src/utils/testing/mockUsers.tsx [hunk 4]
```diff
@@ -20,6 +20,25 @@ export const mockStateUser: UserContextShape = {
...
+export const mockStateUserNoReports: UserContextShape = {
+  user: {
+    userRole: UserRoles.STATE_USER,
+    email: "stateuser@test.com",
+    given_name: "Thelonious",
+    family_name: "States",
+    full_name: "Thelonious States",
+    state: "MN",
+    userIsEndUser: true,
+    userReports: undefined,
```
Comment (braxex) [inline services/ui-src/src/utils/testing/mockUsers.tsx:41]: Question: Do we need to also test for if a user has selected no reports when they provision their role? Idk if state users will have undefined (all other user types definitely will though), but i think they might have the IDM reports attribute equal to an empty string `""` if they didn't select any reports. We should check on this/test for it I think.

**Review by braxex — COMMENTED** [review]
Left some comments and questions re: potential edge cases but didn't have time to explicitly test or go through all the tests and logic.

File: services/app-api/utils/auth/authorization.ts [hunk 5]
```diff
@@ -92,3 +93,21 @@ export const hasPermissions = (
 
   return isAllowed;
 };
+
+export const hasAccess = (event: APIGatewayProxyEvent, reportType: string) => {
```
Comment (BearHanded) [inline services/app-api/utils/auth/authorization.ts:97]: The naming of this method is pretty interchangeable with hasPermissions() at a glance. Maybe hasReportAccess() like the var name you used, or updating this one to hasReport, and the other hasRole?

**Review by BearHanded — CHANGES_REQUESTED** [review]

**Review by keeysnc — COMMENTED** [review]
Hey @karla-vm Im getting an error message when trying to entering mcpar online.
Here's the error:
Screenshot 2023-05-18 at 11 49 47 AM

**Comment by braxex** [comment]
> @karla-vm Did some digging and I think this error is on my end. I can see the 403 permission error showing which seems like its intended for the `staterep` user.

@karla-vm @keeysnc Looks like a bug to me. State reps should be able to fetch, read, write, update, etc. reports.

File: services/ui-src/src/components/cards/TemplateCard.tsx (hunk 2 above)
Comment (karla-vm) [reply services/ui-src/src/components/cards/TemplateCard.tsx:70]: Originally, the top-level / incoming `isDisabled` prop was set to `false` and only used for this particular download button, so I assumed users need to be able to download the template regardless of their access.

File: services/ui-src/src/utils/testing/mockUsers.tsx (hunk 4 above)
Comment (karla-vm) [reply services/ui-src/src/utils/testing/mockUsers.tsx:41]: good call, I changed this unit test to have an array of an empty string, just for consistency across state user roles.

**Review by BearHanded — APPROVED** [review]

**Review by keeysnc — APPROVED** [review]
Nice! Works for me now.

**Review by braxex — APPROVED** [review]

**Review by BearHanded — APPROVED** [review]
